# PROJETO 2: BANCO DE DADOS

### Tabela de Comando SQL

| **Comando SQL**        | **Descrição**                                                                 | **Exemplo de Uso**                                                                                     |
|------------------------|-------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------|
| `CREATE TABLE`          | Cria uma nova tabela no banco de dados.                                       | ```CREATE TABLE Alunos (id INTEGER PRIMARY KEY, nome TEXT);```                                           |
| `INSERT INTO`           | Insere novos registros em uma tabela.                                         | ```INSERT INTO Alunos (id, nome) VALUES (1, 'João');```                                                  |
| `SELECT`                | Seleciona dados de uma ou mais tabelas.                                       | ```SELECT * FROM Alunos;```                                                                             |
| `JOIN`                  | Junta dados de duas ou mais tabelas com base em uma condição.                 | ```SELECT Alunos.nome, Cursos.nome FROM Alunos JOIN Cursos ON Alunos.id_curso = Cursos.id;```             |
| `WHERE`                 | Filtra os resultados de uma consulta com base em uma condição.                | ```SELECT * FROM Alunos WHERE idade > 18;```                                                             |
| `GROUP BY`              | Agrupa resultados de uma consulta em categorias.                             | ```SELECT curso, COUNT(*) FROM Alunos GROUP BY curso;```                                                 |
| `HAVING`                | Filtra grupos após o uso de `GROUP BY`.                                       | ```SELECT curso, COUNT(*) FROM Alunos GROUP BY curso HAVING COUNT(*) > 5;```                             |
| `ORDER BY`              | Ordena os resultados de uma consulta.                                         | ```SELECT * FROM Alunos ORDER BY nome ASC;```                                                            |
| `UPDATE`                | Atualiza registros existentes em uma tabela.                                  | ```UPDATE Alunos SET idade = 20 WHERE id = 1;```                                                         |
| `DELETE`                | Remove registros de uma tabela.                                               | ```DELETE FROM Alunos WHERE id = 1;```                                                                   |
| `ALTER TABLE`           | Modifica a estrutura de uma tabela.                                           | ```ALTER TABLE Alunos ADD COLUMN email TEXT;```                                                          |
| `DROP TABLE`            | Remove uma tabela do banco de dados.                                          | ```DROP TABLE Alunos;```                                                                                 |
| `INNER JOIN`            | Retorna apenas os registros que possuem correspondências em ambas as tabelas. | ```SELECT Alunos.nome, Cursos.nome FROM Alunos INNER JOIN Cursos ON Alunos.id_curso = Cursos.id;```       |
| `LEFT JOIN`             | Retorna todos os registros da tabela à esquerda e os correspondentes à direita.| ```SELECT Alunos.nome, Cursos.nome FROM Alunos LEFT JOIN Cursos ON Alunos.id_curso = Cursos.id;```        |
| `RIGHT JOIN`            | Retorna todos os registros da tabela à direita e os correspondentes à esquerda.| ```SELECT Alunos.nome, Cursos.nome FROM Alunos RIGHT JOIN Cursos ON Alunos.id_curso = Cursos.id;```       |
| `FULL OUTER JOIN`       | Retorna todos os registros quando há correspondência em uma das tabelas.      | ```SELECT Alunos.nome, Cursos.nome FROM Alunos FULL OUTER JOIN Cursos ON Alunos.id_curso = Cursos.id;```  |
| `DISTINCT`              | Retorna valores distintos (não duplicados) de uma coluna.                     | ```SELECT DISTINCT curso FROM Alunos;```                                                                 |
| `COUNT`                 | Conta o número de registros que correspondem a uma consulta.                  | ```SELECT COUNT(*) FROM Alunos;```                                                                       |
| `AVG`                   | Calcula a média de uma coluna numérica.                                       | ```SELECT AVG(idade) FROM Alunos;```                                                                     |
| `SUM`                   | Calcula a soma de uma coluna numérica.                                        | ```SELECT SUM(salario) FROM Funcionarios;```                                                             |
| `MAX`                   | Retorna o valor máximo de uma coluna.                                         | ```SELECT MAX(salario) FROM Funcionarios;```                                                             |
| `MIN`                   | Retorna o valor mínimo de uma coluna.                                         | ```SELECT MIN(salario) FROM Funcionarios;```                                                             |
| `CREATE INDEX`          | Cria um índice para acelerar consultas.                                       | ```CREATE INDEX idx_nome ON Alunos (nome);```                                                            |
| `DROP INDEX`            | Remove um índice.                                                             | ```DROP INDEX idx_nome;```                                                                               |


### Cenário: Análise da Jornada e Experiência do Cliente

#### Descrição do Cenário:


Uma rede de hotéis está interessada em entender melhor a jornada dos seus clientes no site, bem como a experiência de reseva e estadia. A empresa tem informações sobre:

  - **Navegação:** O cliente acessa o site e visualiza diferentes páginas.

  - **Reservas:** O cliente faz uma reserva de quarto em um hotel.

  - **Avaliações:** O cliente avalia sua experiência após a estadia.

A rede quer analisar como diferentes perfis de clientes avaliam a experiência de estadia, comparando esses resultados com todo ofunil até chegar a estadia, desde a navegação até a avaliação do hotel.

#### Objetivo:

  - Entender a jornada do cliente desde a navegação no site até a reserva e sua experiência final.

  - Perguntas para análise:
  1. Qual o tempo médio que os clientes gastam no site antes de fazer uma reserva?

  2. Qual o percentual de clientes que realizam reservas após a navegação?

3. Existe alguma correlação entre o tempo gasto no site e a avaliação dada?

### O que você precisa fazer no Projeto 2

1. Inserir os dados dos arquivos csv em um banco de dados SQLite.

2. Realizar consultas para responder às perguntas de análise.

3. Salvar as consultas como tabela no banco de dados.

DICA:

Para verificar todas as tabelas que existem em um banco de dados SQLite, você pode rodar O comando abaixo.

Isso exibirá uma lista com todas as tabelas do banco de dados, permitindo verificar se as tabelas  foram criadas corretamente.

In [ ]:
#cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
#tabelas = cursor.fetchall()
#print(tabelas)

Se quiser excluir bases do banco de dados use o comando abaixo:

In [ ]:
#cursor.execute('DROP TABLE IF EXISTS nome_da_tabela')

###Preparando o ambiente

In [ ]:
# Montando o Google Drive
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [ ]:
# direciona para o caminho desejado
%cd /content/drive/MyDrive/PRETALAB/Banco\ de\ Dados/Projeto2/

/content/drive/MyDrive/PRETALAB/Projeto 2


In [ ]:
#Este módulo fornece uma maneira simples de usar funcionalidades que são dependentes de sistema operacional
import os

# os.getcwd() retorna o caminho absoluto do diretório de trabalho atual
print(os.getcwd())

/content/drive/MyDrive/PRETALAB/Projeto 2


- Criando um banco de dados chamado 'hotel'

In [ ]:
#Importar as bibliotecas necessárias
import sqlite3 as sql
import csv

In [ ]:
 #Conectar (ou criar) um banco de dados
conexao = sql.connect('hotel.db')
cursor = conexao.cursor()

In [ ]:
#Excluir tabelas caso ja existam
cursor.execute('DROP TABLE IF EXISTS clientes')
cursor.execute('DROP TABLE IF EXISTS avaliacoes')
cursor.execute('DROP TABLE IF EXISTS navegacao')
cursor.execute('DROP TABLE IF EXISTS reservas')
cursor.execute('DROP TABLE IF EXISTS avaliacao_clientes_hotel')

- Criando tabelas

In [ ]:
# Função para carregar dados de um arquivo CSV para uma tabela SQL:
def salvar_csv_no_banco(nome_arquivo_csv, nome_tabela, conexao):
    with open(nome_arquivo_csv, 'r') as file:
        reader = csv.reader(file)
        colunas = next(reader)  # Pular a linha de cabeçalho
        dados = [tuple(linha) for linha in reader]
#Criar a tabela no banco de dados:
        cursor = conexao.cursor()
        cursor.execute(f"CREATE TABLE IF NOT EXISTS {nome_tabela} ({', '.join(colunas)})")
#Inserir os dados na tabela:
        placeholders = ', '.join(['?'] * len(colunas))
        for linha in dados:
                    try:
                        cursor.execute(f"INSERT INTO {nome_tabela} VALUES ({placeholders})", linha)
                    except sql.IntegrityError:
                        print(f"Erro: O id_cliente {linha[0]} já existe na tabela.")
                    except Exception as e:
                        print(f"Erro ao inserir dados na tabela {nome_tabela}: {e}")

        conexao.commit()
        print(f"Tabela '{nome_tabela}' criada e dados inseridos com sucesso.")

# Salvar os arquivos CSV no banco de dados
salvar_csv_no_banco('clientes.csv', 'clientes', conexao)
salvar_csv_no_banco('avaliacoes.csv', 'avaliacoes', conexao)
salvar_csv_no_banco('navegacao.csv', 'navegacao', conexao)
salvar_csv_no_banco('reservas.csv', 'reservas', conexao)

Tabela 'clientes' criada e dados inseridos com sucesso.
Tabela 'avaliacoes' criada e dados inseridos com sucesso.
Tabela 'navegacao' criada e dados inseridos com sucesso.
Tabela 'reservas' criada e dados inseridos com sucesso.


In [ ]:
# Consultar todos os dados da tabela "clientes"
cursor.execute("SELECT * FROM clientes")
resultados = cursor.fetchall()

# Exibir os dados recuperados
for linha in resultados:
    print(linha)

('1', 'Frank Watts', 'alyssa06@example.net', '65', 'Feminino')
('2', 'Debbie Smith', 'alison46@example.org', '23', 'Masculino')
('3', 'Autumn Decker', 'marygarner@example.com', '45', 'Feminino')
('4', 'Kyle Diaz', 'craigstephanie@example.com', '63', 'Masculino')
('5', 'Patricia Mitchell DDS', 'pdavis@example.org', '60', 'Feminino')
('6', 'Julie Garcia', 'xhernandez@example.com', '72', 'Masculino')
('7', 'Steven Bryan', 'keith12@example.com', '30', 'Masculino')
('8', 'Jamie Martin', 'davidjohnson@example.net', '36', 'Masculino')
('9', 'Nathan Peterson', 'james93@example.com', '42', 'Feminino')
('10', 'Joseph Moore', 'sandra07@example.org', '69', 'Masculino')
('11', 'Alexa Hunt', 'tknapp@example.com', '39', 'Feminino')
('12', 'Stephen Garrett', 'deborahgrimes@example.org', '63', 'Feminino')
('13', 'Susan Watson', 'danieldominguez@example.org', '27', 'Masculino')
('14', 'Tim Brown', 'bcarter@example.net', '23', 'Feminino')
('15', 'Alexandria Payne', 'amber52@example.com', '93', 'Masculino

In [ ]:
# Consultar todos os dados da tabela "avaliacoes"
cursor.execute("SELECT * FROM avaliacoes")
resultados = cursor.fetchall()

# Exibir os dados recuperados
for linha in resultados:
    print(linha)

('1', '1336', '8', 'Number ago network provide upon the once whether local you break they along response many but not continue pattern few home three behind million sometimes such.')
('2', '2851', '5', 'Sure simply interesting say policy other along evidence edge drive several debate protect play whatever purpose keep budget morning.')
('3', '2259', '2', 'Anyone doctor reflect opportunity nice team of majority any administration beautiful information go air month rise animal sound among beautiful stand cause.')
('4', '4646', '6', 'Early provide TV imagine people son message cell method senior measure great card arrive serious wear upon wrong national various community team west bank company themselves prepare.')
('5', '3626', '3', 'Sister physical money view surface hope whether future likely paper nature new scene teacher outside view weight population official support doctor detail.')
('6', '3250', '7', 'Price nor must five easy present current change bad production case bad order ri

In [ ]:
# Consultar todos os dados da tabela "navegacao"
cursor.execute("SELECT * FROM navegacao")
resultados = cursor.fetchall()

# Exibir os dados recuperados
for linha in resultados:
    print(linha)

('1', '4810', 'busca_hoteis', '29', '2024-02-23')
('2', '1342', 'checkout', '13', '2024-09-06')
('3', '1117', 'checkout', '75', '2023-11-24')
('4', '4017', 'home', '80', '2024-07-15')
('5', '1622', 'busca_hoteis', '47', '2023-12-01')
('6', '4578', 'busca_hoteis', '4', '2023-12-30')
('7', '4840', 'checkout', '51', '2023-10-04')
('8', '21', 'checkout', '68', '2024-06-25')
('9', '1190', 'home', '4', '2024-03-26')
('10', '1706', 'busca_hoteis', '52', '2023-12-07')
('11', '2980', 'detalhe_hotel', '12', '2024-04-12')
('12', '174', 'home', '16', '2024-03-14')
('13', '2261', 'detalhe_hotel', '1', '2024-02-12')
('14', '2327', 'busca_hoteis', '101', '2023-11-12')
('15', '1547', 'detalhe_hotel', '62', '2023-11-09')
('16', '4491', 'busca_hoteis', '19', '2023-10-06')
('17', '2848', 'busca_hoteis', '82', '2024-09-11')
('18', '1520', 'home', '32', '2024-08-29')
('19', '4618', 'home', '86', '2024-07-03')
('20', '686', 'home', '77', '2023-10-01')
('21', '4612', 'busca_hoteis', '68', '2024-06-06')
('22'

In [ ]:
# Consultar todos os dados da tabela "reservas"
cursor.execute("SELECT * FROM reservas")
resultados = cursor.fetchall()

# Exibir os dados recuperados
for linha in resultados:
    print(linha)

('1', '4762', '2024-06-28', '2024-07-19', '9738.74')
('2', '1832', '2024-04-22', '2024-05-20', '5462.52')
('3', '2002', '2024-07-06', '2024-08-12', '2529.75')
('4', '1865', '2024-03-17', '2024-05-22', '1435.9')
('5', '4106', '2024-05-28', '2024-07-05', '2713.77')
('6', '3005', '2024-04-08', '2024-08-29', '3726.64')
('7', '1066', '2024-06-24', '2024-07-17', '7851.19')
('8', '2335', '2024-02-08', '2024-06-04', '6487.5')
('9', '1758', '2023-12-11', '2024-08-07', '9699.13')
('10', '4848', '2023-10-17', '2024-08-27', '7891.05')
('11', '661', '2023-10-05', '2023-11-08', '560.01')
('12', '3455', '2024-01-15', '2024-02-16', '3879.52')
('13', '1621', '2024-06-04', '2024-07-25', '5807.32')
('14', '3066', '2024-06-24', '2024-06-27', '871.86')
('15', '967', '2024-09-16', '2024-09-30', '9311.48')
('16', '2780', '2024-08-13', '2024-09-30', '7726.88')
('17', '2869', '2023-12-26', '2024-03-29', '9947.23')
('18', '2645', '2024-04-19', '2024-09-02', '6896.95')
('19', '3390', '2024-01-26', '2024-04-10', 

In [ ]:
# Salvar alterações
conexao.commit()
print("Dados inseridos com sucesso no banco de dados!")

Dados inseridos com sucesso no banco de dados!


In [ ]:
# Executar o JOIN e criar uma nova tabela
cursor.execute('''
CREATE TABLE IF NOT EXISTS avaliacao_clientes_hotel AS
SELECT *
FROM clientes
JOIN avaliacoes ON clientes.id_cliente = avaliacoes.id_cliente
JOIN reservas ON clientes.id_cliente = reservas.id_cliente
JOIN navegacao ON clientes.id_cliente = navegacao.id_cliente
''')

print("Nova tabela 'avaliacao_clientes_hotel' criada e dados inseridos com sucesso.")

Nova tabela 'avaliacao_clientes_hotel' criada e dados inseridos com sucesso.


In [ ]:
# Consultar os nomes das colunas da tabela "avaliacao_clientes_hotel"
cursor.execute("PRAGMA table_info(avaliacao_clientes_hotel)")
colunas = cursor.fetchall()

# Exibir os nomes das colunas
for coluna in colunas:
    print(coluna[1])  # O nome da coluna está no índice 1

id_cliente
nome
email
idade
genero
id_avaliacao
id_cliente:1
nota
comentario
id_reserva
id_cliente:2
data_checkin
data_checkout
valor_total
id_navegacao
id_cliente:3
pagina
tempo_gasto
data_acesso


In [ ]:
# Consultar todos os dados da tabela "avaliacao_clientes_hotel"
cursor.execute("SELECT * FROM avaliacao_clientes_hotel")
resultados = cursor.fetchall()

# Exibir os dados recuperados
for linha in resultados:
    print(linha)

('2', 'Debbie Smith', 'alison46@example.org', '23', 'Masculino', '763', '2', '2', 'Training still able matter view prepare senior heart child collection month low threat picture however local know a seek beyond.', '1579', '2', '2024-05-19', '2024-07-12', '4421.72', '2737', '2', 'busca_hoteis', '36', '2024-04-13')
('3', 'Autumn Decker', 'marygarner@example.com', '45', 'Feminino', '1757', '3', '1', 'Could with program unit soldier someone whose war first later campaign Republican his these politics customer trouble traditional meet simply step when carry far lawyer.', '2364', '3', '2024-05-27', '2024-09-22', '9706.13', '1017', '3', 'busca_hoteis', '103', '2024-07-04')
('3', 'Autumn Decker', 'marygarner@example.com', '45', 'Feminino', '1757', '3', '1', 'Could with program unit soldier someone whose war first later campaign Republican his these politics customer trouble traditional meet simply step when carry far lawyer.', '332', '3', '2024-05-25', '2024-06-05', '2004.68', '1017', '3', 'bu

In [ ]:
#Verificar as tabelas existentes no banco de dados 'hotel'
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tabelas = cursor.fetchall()
print(tabelas)

[('correlacao_tempo_gasto_media_avaliacoes',), ('tempo_medio_reservas',), ('percentual_reservas',), ('clientes',), ('avaliacoes',), ('navegacao',), ('reservas',), ('avaliacao_clientes_hotel',)]


## Perguntas para análise:

### 1. 📊✨ Qual o tempo médio que os clientes gastam no site antes de fazer uma reserva?

In [ ]:
# Consulta SQL para verificar o tempo médio em dias que os clientes entram no site e fazem a reserva
cursor.execute('''
CREATE TABLE IF NOT EXISTS tempo_medio_reservas AS
SELECT
    AVG(JULIANDAY(r.data_checkin) - JULIANDAY(n.data_acesso)) AS tempo_medio_dias
FROM
    reservas r
JOIN
    navegacao n ON r.id_cliente = n.id_cliente
WHERE
    n.data_acesso < r.data_checkin;
''')

conexao.commit()  # Confirmar as mudanças
print("Tabela 'tempo_medio_reservas' criada com sucesso.")

Tabela 'tempo_medio_reservas' criada com sucesso.


In [ ]:
# Consultar os dados da tabela criada
cursor.execute('''
SELECT * FROM tempo_medio_reservas;
''')

resultados = cursor.fetchall()
tempo_medio_dias = resultados[0][0]

# Exibir o resultado
print(f"O tempo médio que os clientes gastam no site antes de fazer uma reserva é de aproximadamente {tempo_medio_dias:.2f} dias.")

O tempo médio que os clientes gastam no site antes de fazer uma reserva é de aproximadamente 123.08 dias.


**Com base na nossa análise, podemos concluir que os clientes levam, em média, 123 dias desde o primeiro acesso ao site até realizarem o check-in no hotel. Isso sugere um longo período de consideração e planejamento antes de finalizarem a reserva.**

### 2. 📊🔍 Qual o percentual de clientes que realizam reservas após a navegação?

In [ ]:
# Percentual de clientes que realizaram reservas após a navegação:
cursor.execute('''
CREATE TABLE IF NOT EXISTS percentual_reservas AS
SELECT
    (COUNT(DISTINCT r.id_cliente) * 100.0 / (SELECT COUNT(DISTINCT id_cliente) FROM navegacao)) AS percentual_clientes_reservas
FROM
    reservas r
JOIN
    navegacao n ON r.id_cliente = n.id_cliente
WHERE
    n.data_acesso < r.data_checkin;
''')

conexao.commit()  # Confirmar as mudanças
print("Tabela 'percentual_reservas' criada com sucesso.")

Tabela 'percentual_reservas' criada com sucesso.


In [ ]:
# Consultar os dados da tabela criada
cursor.execute('''
SELECT * FROM percentual_reservas;
''')

resultados = cursor.fetchall()
percentual_clientes_reservas = resultados[0][0]

# Exibir o resultado
print(f"O percentual de clientes que realizaram reservas após a navegação é de aproximadamente {percentual_clientes_reservas:.2f}%.")

O percentual de clientes que realizaram reservas após a navegação é de aproximadamente 42.12%.


**De acordo com o cálculo realizado, aproximadamente 42,12% dos clientes que navegam no site acabam concluindo uma reserva. Esse dado é bastante revelador e pode ajudar a entender melhor o comportamento dos usuários.**

### 3. Existe alguma correlação entre o tempo gasto no site e a avaliação dada?

In [ ]:
# Consulta SQL para contar total de tempo gasto e calcular a média das notas de avaliação por cliente:
cursor.execute('''
SELECT
    clientes.id_cliente, clientes.nome, COUNT(navegacao.tempo_gasto) AS total_tempo_gasto, AVG(avaliacoes.nota) AS media_avaliacao
FROM
    clientes
JOIN
    navegacao ON clientes.id_cliente = navegacao.id_cliente
JOIN
    avaliacoes ON clientes.id_cliente = avaliacoes.id_cliente
WHERE
    navegacao.id_cliente = avaliacoes.id_cliente
GROUP BY
    clientes.nome;
''')

# Exibir resultado
resultados = cursor.fetchall()
for resultado in resultados:
    print(f"Cliente: {resultado[1]}, Total de tempo gasto: {resultado[2]}, Média de Avaliação: {resultado[3]:.2f}")

Cliente: Aaron Clark, Total de tempo gasto: 1, Média de Avaliação: 8.00
Cliente: Aaron Fisher, Total de tempo gasto: 4, Média de Avaliação: 5.50
Cliente: Aaron Foster, Total de tempo gasto: 2, Média de Avaliação: 4.50
Cliente: Aaron Graves, Total de tempo gasto: 2, Média de Avaliação: 7.00
Cliente: Aaron Mcclure, Total de tempo gasto: 1, Média de Avaliação: 3.00
Cliente: Aaron Orozco, Total de tempo gasto: 2, Média de Avaliação: 3.00
Cliente: Abigail Berry, Total de tempo gasto: 2, Média de Avaliação: 6.00
Cliente: Abigail Gonzales, Total de tempo gasto: 2, Média de Avaliação: 1.00
Cliente: Abigail Jones, Total de tempo gasto: 2, Média de Avaliação: 3.00
Cliente: Adam Allen, Total de tempo gasto: 1, Média de Avaliação: 7.00
Cliente: Adam Bowen, Total de tempo gasto: 1, Média de Avaliação: 1.00
Cliente: Adam Fischer, Total de tempo gasto: 1, Média de Avaliação: 10.00
Cliente: Adam Foster, Total de tempo gasto: 2, Média de Avaliação: 4.00
Cliente: Adam Johnson, Total de tempo gasto: 3, M

In [ ]:
# Criar a tabela para armazenar a correlação entre o total de tempo gasto e calcular a média das notas de avaliação:
cursor.execute('''
CREATE TABLE IF NOT EXISTS correlacao_tempo_gasto_media_avaliacoes (
    id_cliente INTEGER,
    nome TEXT,
    total_tempo_gasto INTEGER,
    media_avaliacao REAL,
    PRIMARY KEY(id_cliente)
)
''')

# Inserir os resultados na tabela correlacao_tempo_gasto_media_avaliacoes
resultados = cursor.fetchall()
cursor.executemany('''
INSERT OR REPLACE INTO correlacao_tempo_gasto_media_avaliacoes (id_cliente, nome, total_tempo_gasto, media_avaliacao)
VALUES (?, ?, ?, ?)
''', resultados)

print("Nova tabela 'correlacao_tempo_gasto_media_avaliacoes' criada e dados inseridos com sucesso.")

Nova tabela 'correlacao_tempo_gasto_media_avaliacoes' criada e dados inseridos com sucesso.


In [ ]:
# Salvar as alterações e fechar a conexão
conexao.commit()
conexao.close()